### PR TITLE
Angular. Components. Directives & Pipes task

### PR DESCRIPTION
**Pull request by iamsixthousand**
# Angular. Components. Directives & Pipes task

## Task requirements
The main goal of this task is to implement the main page of the YouTube client app. You are to enrich previously generated components with necessary functionality and markup.

### Functional requirements
- Once a user is entered the app, The only **Header** component should be shown.
For now, you can enter whatever value which won't influence your search result.
Please note that the **Sorting criteria block** should be hidden
- By submitting the search form, the **Search results block** appears.
- Using the [response example](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/response.json), render cards with predefined data
    You can store the mocked data in the **Search results** component
    - The bottom border represents the publication date status. Create a directive to achieve the result
        - If a publication date is less than a month, set border background to green
        - If a publication date is less than 7 days, set border background to blue
        - If a publication date is more than 6 months, set border background to red
- By clicking on the **Settings button**, the **Filtering criteria block** should be toggled
    - It should be possible to sort search results by date or count of views
    - Sorting should work both in the direction of decreasing values and in the direction of increasing values
    - Using a pipe, filter search results by value that a user types in the input

#### TODO
- [x] Main layout is implemented;
- [x] Implemented a fake search functionality: by submitting the form on the **Header** component, search results are shown;
- [x] The **Search result item** component contains all the necessary data;
- [x] The **Filtering criteria block** toggle functionality is implemented;
- [x] The colorful border under the **Search result item** is implemented;
- [x] Search result sorting is implemented (Should work both in the direction of decreasing values, and in the direction of increasing values);
- [x] Search result filtering by key words is implemented;

### CODE REVIEW

File: src/styles.scss [hunk 1]
```diff
@@ -1 +1,15 @@
-/* You can add global styles to this file, and also import other style files */
+@import "./style/variables.scss";
+@import "./style/mixins.scss";
+@import "./style/functions.scss";
+
+:root {
+  font-size: 16px;
```
Comment (AnatoliRub) [inline src/styles.scss]: please change px to rem

File: src/app/core/components/header/logo/logo.component.scss [hunk 2]
```diff
@@ -0,0 +1,8 @@
+@import '/src/styles.scss';
+
+.logo {
+    padding-right: 10px;
```
Comment (AnatoliRub) [inline src/app/core/components/header/logo/logo.component.scss]: please fix it to rem

File: src/app/core/components/header/search/search.component.scss [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        }
+    }
+
+    &__submit {
+        background-color: var(--secondary-bg);
+        color: var(--white-color);
+        width: 5rem;
+        height: 27px;
+        border: none;
+        border-radius: 0 4px 4px 0;
```
Comment (AnatoliRub) [inline src/app/core/components/header/search/search.component.scss]: please fix it to rem

File: src/app/core/components/header/search-settings/search-settings.component.ts [hunk 4]
```diff
@@ -1,10 +1,17 @@
 /* eslint-disable @typescript-eslint/no-useless-constructor */
 /* eslint-disable @typescript-eslint/no-empty-function */
```
Comment (AnatoliRub) [inline src/app/core/components/header/search-settings/search-settings.component.ts]: please remove comments

File: src/app/core/components/screen/card/card-stats/comments/comments.component.html [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<div class="card__stats-comments-container">
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/comments/comments.component.html:1]: plese fix structure according prettier rules

File: src/app/core/components/screen/card/card-stats/dislikes/dislikes.component.html [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<div class="card__stats-dislikes-container">
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/dislikes/dislikes.component.html:1]: please fix structure according to prettier rules

File: src/app/core/components/screen/card/card-stats/likes/likes.component.html [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<div class="card__stats-likes-container">
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/likes/likes.component.html:1]: please fix structure

File: src/app/core/components/screen/card/card-stats/views/views.component.html [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<div class="card__stats-views-container">
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/views/views.component.html:1]: please fix structure

File: src/app/pipes/shortumber.pipe.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'shortNumber',
+})
+export class ShortNumberPipe implements PipeTransform {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (AnatoliRub) [inline src/app/pipes/shortumber.pipe.ts]: please remove comment and delete unused args param if it possible

File: src/app/pipes/shortumber.pipe.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  transform(string: string | undefined, args?: any): string | null {
```
Comment (AnatoliRub) [inline src/app/pipes/shortumber.pipe.ts]: I think we could simplify implementation transform method 
Possible implementation example:
const formatCounts = (counts) => {
const sumDigits = counts.split('').length;
if(sumDigits >= 10) return counts.slice(0, -9) + 'B';
if(sumDigits >= 7) return counts.slice(0, -6) + 'M';
if(sumDigits >= 4) return counts.slice(0, -3) + 'K';
return counts;
} 
It seems to me This implementation more readble.

File: src/app/app.component.ts [hunk 11]
```diff
@@ -1,10 +1,25 @@
...
+
+  public $isFilterVisible: Observable<boolean> =
+    this.searchService.showFilterFlag;
+
+  showFilter: boolean = false;
+
+  posts: PostItem[] = data.items;
+
+  ngOnInit(): void {
+    console.log(this.showFilter);
```
Comment (AnatoliRub) [inline src/app/app.component.ts]: please remove console.log and NgOnInit

File: src/app/core/components/screen/card/card.component.scss [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+.card {
+    background-color: var(--main-bg);
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    justify-content: baseline;
+    width: 16rem;
+    height: 20rem;
+    border-radius: 0.5rem;
+    box-shadow: var(--shadow-black) 1.95px 1.95px 2.6px;
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card.component.scss]: please fix pix to rem

File: src/app/core/components/screen/card/card-stats/card-stats.component.html [hunk 13]
```diff
@@ -1 +1,18 @@
-<p>card-stats works!</p>
+<div class="card__stats">
+    <div class="card__stats-element">
+        <app-card-stats-views></app-card-stats-views>
+        {{stats?.viewCount | shortNumber}}
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/card-stats.component.html]: please wrap value of conts into tag

File: src/app/core/components/screen/card/card-stats/comments/comments.component.scss [hunk 14]
```diff
@@ -0,0 +1,5 @@
+.card__stats-comments-container {
+    display: flex;
```
Comment (AnatoliRub) [inline src/app/core/components/screen/card/card-stats/comments/comments.component.scss]: We could to add mixin with these rules and reusable it into dislikes, likes and views components

File: src/app/core/components/screen/screen.component.ts [hunk 15]
```diff
@@ -1,10 +1,14 @@
 /* eslint-disable @typescript-eslint/no-useless-constructor */
```
Comment (AnatoliRub) [inline src/app/core/components/screen/screen.component.ts]: please remove comments

File: src/app/pipes/shortumber.pipe.ts (hunk 10 above)
Comment (StanevichIvan) [reply src/app/pipes/shortumber.pipe.ts]: @AnatoliRub  please use 
image

Comment (AnatoliRub) [reply src/app/pipes/shortumber.pipe.ts]:
```suggestion
  const formatCounts = (counts) => {
const sumDigits = counts.split('').length;
if(sumDigits >= 10) return counts.slice(0, -9) + 'B';
if(sumDigits >= 7) return counts.slice(0, -6) + 'M';
if(sumDigits >= 4) return counts.slice(0, -3) + 'K';
return counts;
}
```

File: src/styles.scss (hunk 1 above)
Comment (iamsixthousand) [reply src/styles.scss]: it is a root font-size

Comment (AnatoliRub) [reply src/styles.scss]: I'm sorry.^.^,

File: src/app/app-routing.module.ts [hunk 16]
```diff
@@ -1,11 +1,37 @@
...
+    component: SearchResultPageComponent,
+  },
+  {
+    path: RouterPath.Register,
+    component: RegistrationPageComponent,
+  },
+  {
+    path: RouterPath.Redirect,
+    redirectTo: RouterPath.Main,
+    // pathMatch: 'full',
```
Comment (AnatoliRub) [inline src/app/app-routing.module.ts]: Please delete all comments

File: src/app/core/services/search-service.service.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+export class SearchServiceService {
+  postData: PostItem[] = data.items;
+
+  #showFilter: boolean = false;
+
+  #showFilterFlag: BehaviorSubject<boolean> = new BehaviorSubject(
+    this.#showFilter
+  );
+
+  showFilterFlag: Observable<boolean> = this.#showFilterFlag.asObservable();
```
Comment (AnatoliRub) [inline src/app/core/services/search-service.service.ts]: Please add $

File: src/app/app.component.ts [hunk 18]
```diff
@@ -1,10 +1,17 @@
 import { Component } from '@angular/core';
+import { Observable } from 'rxjs';
+import { SearchServiceService } from './core/services/search-service.service';
 
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss'],
 })
 export class AppComponent {
```
Comment (AnatoliRub) [inline src/app/app.component.ts:19]: please check the order. https://palantir.github.io/tslint/rules/member-ordering/

File: src/app/core/components/header/login-info/login-info.component.ts [hunk 19]
```diff
@@ -1,10 +1,19 @@
...
+import { faUserCircle } from '@fortawesome/free-solid-svg-icons';
+import { RouterPath } from 'src/app/routes.enum';
 
 @Component({
   selector: 'app-login-info',
   templateUrl: './login-info.component.html',
   styleUrls: ['./login-info.component.scss'],
 })
-export class LoginInfoComponent {}
+export class LoginInfoComponent {
```
Comment (AnatoliRub) [inline src/app/core/components/header/login-info/login-info.component.ts:11]: please check the functions order. https://palantir.github.io/tslint/rules/member-ordering/

File: src/app/core/components/header/search/search.component.ts [hunk 20]
```diff
@@ -1,10 +1,26 @@
...
+import { RouterPath } from 'src/app/routes.enum';
+import stringConstants from 'src/constants/constants';
 
 @Component({
   selector: 'app-search',
   templateUrl: './search.component.html',
   styleUrls: ['./search.component.scss'],
 })
-export class SearchComponent {}
+export class SearchComponent {
```
Comment (AnatoliRub) [inline src/app/core/components/header/search/search.component.ts:18]: please check the functions order. https://palantir.github.io/tslint/rules/member-ordering/

File: src/app/core/components/screen/pages/search-result-page/search-result-page.component.ts [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+import { Observable } from 'rxjs';
+import { SearchServiceService } from 'src/app/core/services/search-service.service';
+import { PostItem } from 'src/interfaces/youtube';
+
+@Component({
+  selector: 'app-search-result-page',
+  templateUrl: './search-result-page.component.html',
+  styleUrls: ['./search-result-page.component.scss'],
+})
+export class SearchResultPageComponent {
```
Comment (AnatoliRub) [inline src/app/core/components/screen/pages/search-result-page/search-result-page.component.ts:12]: please check the functions order. https://palantir.github.io/tslint/rules/member-ordering/

File: src/app/app.component.ts [hunk 22]
```diff
@@ -1,10 +1,17 @@
...
-  title = 'youtube-client-angular';
+  constructor(private readonly searchService: SearchServiceService) {}
+
+  public $isFilterVisible: Observable<boolean> =
+    this.searchService.showFilterFlag;
+
+  showFilter: boolean = false;
```
Comment (StanevichIvan) [inline src/app/app.component.ts]: Do we use this variable somewhere?

File: src/app/core/components/filter/filter.component.scss [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+    align-self: center;
+    justify-content: center;
+  }
+
+  &__itemlist {
+    padding: 0 0.8rem;
+    list-style-type: none;
+    display: flex;
+    flex-direction: row;
+    gap: 0.8rem;
```
Comment (StanevichIvan) [inline src/app/core/components/filter/filter.component.scss]: might pxToRem

File: src/app/core/components/header/header.component.html [hunk 24]
```diff
@@ -1 +1,8 @@
-<p>header works!</p>
+<div class="header">
+    <app-logo></app-logo>
+    <app-search></app-search>
+    <app-search-settings></app-search-settings>
+    <app-login-info></app-login-info>
+</div>
+
+
```
Comment (StanevichIvan) [inline src/app/core/components/header/header.component.html]: let's add html to prettier config

File: src/app/core/components/header/login-info/login-info.component.scss [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+    opacity: 0.8;
+
+    &:hover {
+        opacity: 1;
+    }
+
+    &:active {
+        color: var(--secondary-bg);
+    }
+}
```
Comment (StanevichIvan) [inline src/app/core/components/header/login-info/login-info.component.scss]: please add scss to prettier config

File: src/app/core/components/header/logo/logo.component.html [hunk 26]
```diff
@@ -1 +1,5 @@
-<p>logo works!</p>
+<div class="logo" (click)="toTheMainPage()">
+    <svg xmlns="http://www.w3.org/2000/svg" width="50" height="50" fill="currentColor" viewBox="0 0 16 16">
```
Comment (StanevichIvan) [inline src/app/core/components/header/logo/logo.component.html]: Might we could extract sag into separate component

File: src/app/core/components/header/logo/logo.component.scss [hunk 27]
```diff
@@ -0,0 +1,9 @@
+@import '/src/styles.scss';
+
+.logo {
+    cursor: pointer;
+    padding-right: rem(10);
+    display: flex;
+    align-items: center;
+    justify-content: center;
+}
```
Comment (StanevichIvan) [inline src/app/core/components/header/logo/logo.component.scss]: please check config and add empty line

File: src/app/core/components/header/search-settings/search-settings.component.scss [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+  background: var(--red-color);
+  border-radius: rem(5);
+  opacity: 0.8;
+  transition: 1s;
+  cursor: pointer;
+
+  &:hover {
+    .top {
+      &::before {
+        transition: 1s;
```
Comment (StanevichIvan) [inline src/app/core/components/header/search-settings/search-settings.component.scss]: let's create a variable for it

File: src/app/core/components/header/search-settings/search-settings.component.scss [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+    position: absolute;
+    top: -(rem(1));
+    right: rem(2);
+    background-color: var(--white-color);
+    border-radius: rem(1.75);
+    width: rem(3.5);
+    height: rem(3.5);
+    transition: 1s;
+  }
+}
```
Comment (StanevichIvan) [inline src/app/core/components/header/search-settings/search-settings.component.scss]: please add empty line

File: src/app/core/components/screen/card/card-info-button/card-info-button.component.scss [hunk 30]
```diff
@@ -0,0 +1,22 @@
+@import "/src/styles.scss";
+
+.card__more-button-wrapper {
+  height: 5rem;
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-info-button/card-info-button.component.scss]: please add pxToRem

File: src/app/core/components/screen/card/card-info-button/card-info-button.component.scss [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+  align-items: center;
+  justify-content: center;
+}
+
+.card__more-button {
+    background-color: var(--red-color);
+    color: var(--white-color);
+    cursor: pointer;
+    border: none;
+    border-radius: 0.25rem;
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-info-button/card-info-button.component.scss]: please add pxTORem or stick to px

File: src/app/core/components/screen/card/card-info-button/card-info-button.component.ts [hunk 32]
```diff
@@ -1,10 +1,11 @@
...
+import stringConstants from 'src/constants/constants';
 
 @Component({
   selector: 'app-card-info-button',
   templateUrl: './card-info-button.component.html',
   styleUrls: ['./card-info-button.component.scss'],
 })
-export class CardInfoButtonComponent {}
+export class CardInfoButtonComponent {
+  buttonText = stringConstants.moreButtonText;
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-info-button/card-info-button.component.ts]: Might we could inject constants file?

File: src/app/core/components/screen/card/card-stats/card-stats.component.html [hunk 33]
```diff
@@ -1 +1,18 @@
...
+    </div>
+    <div class="card__stats-element">
+        <app-card-stats-dislikes></app-card-stats-dislikes>
+        <div>{{stats?.dislikeCount | shortNumber}}</div>
+    </div>
+    <div class="card__stats-element">
+        <app-card-stats-comments></app-card-stats-comments>
+        <div>{{stats?.commentCount | shortNumber}}</div>
+    </div>
+</div>
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-stats/card-stats.component.html]: please add empty line

File: src/app/core/components/screen/card/card-stats/card-stats.component.scss [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+    font-size: 0.8rem;
+
+    &-element {
+        display: flex;
+        flex-direction: row;
+        align-items: center;
+        justify-content: center;
+        gap: 0.2rem;
+    }
+}
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-stats/card-stats.component.scss]: please add empty line

File: src/app/core/components/screen/card/card-stats/card-stats.component.scss [hunk 35]
```diff
@@ -0,0 +1,18 @@
...
+    align-items: center;
+    gap: 1rem;
+    font-size: 0.8rem;
+
+    &-element {
+        display: flex;
+        flex-direction: row;
+        align-items: center;
+        justify-content: center;
+        gap: 0.2rem;
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-stats/card-stats.component.scss]: please add pxToRem

File: src/app/core/components/screen/card/card-stats/likes/likes.component.html [hunk 36]
```diff
@@ -0,0 +1,16 @@
...
+  <svg
```
Comment (StanevichIvan) [inline src/app/core/components/screen/card/card-stats/likes/likes.component.html]: please create icon moon font with current svg or create stand alone component

File: src/app/core/components/screen/components/start/start.component.ts [hunk 37]
```diff
@@ -0,0 +1,15 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-start',
+  templateUrl: './start.component.html',
+  styleUrls: ['./start.component.scss']
+})
+export class StartComponent implements OnInit {
+
+  constructor() { }
```
Comment (StanevichIvan) [inline src/app/core/components/screen/components/start/start.component.ts]: please remove empty functions

File: src/app/core/components/screen/pages/login-page/login-page.component.ts [hunk 38]
```diff
@@ -0,0 +1,15 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-login-page',
+  templateUrl: './login-page.component.html',
+  styleUrls: ['./login-page.component.scss']
+})
+export class LoginPageComponent implements OnInit {
+
+  constructor() { }
```
Comment (StanevichIvan) [inline src/app/core/components/screen/pages/login-page/login-page.component.ts]: please remove empty functions

File: src/app/core/components/screen/pages/search-result-page/search-result-page.component.scss [hunk 39]
```diff
@@ -0,0 +1,19 @@
+@import "/src/styles.scss";
+
+.search-result-page {
+  display: grid;
+  grid-template-columns: 1fr 1fr 1fr 1fr;
```
Comment (StanevichIvan) [inline src/app/core/components/screen/pages/search-result-page/search-result-page.component.scss]: please use repeat

File: src/app/core/components/screen/screen.component.scss [hunk 40]
```diff
@@ -0,0 +1,11 @@
...
+
+.screen {
+    min-height: calc(100vh - var(--page-padding-top) - var(--page-padding-bottom));
+    height: 100%;
+    display: flex;
+    padding: 9rem 0 5rem 0;
+    align-items: flex-start;
+    justify-content: center;
+    background-color: var(--screen-bg);
+}
```
Comment (StanevichIvan) [inline src/app/core/components/screen/screen.component.scss]: empty line is missing

File: src/app/pipes/shortumber.pipe.ts [hunk 41]
```diff
@@ -0,0 +1,30 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'shortNumber',
```
Comment (StanevichIvan) [inline src/app/pipes/shortumber.pipe.ts:4]: please add pure: true

File: src/constants/constants.ts [hunk 42]
```diff
@@ -0,0 +1,7 @@
+const stringConstants = {
+  searchPlaceholderValue: 'Search for anything',
+  searchButtonSubmit: 'Search',
+  moreButtonText: 'more..',
+};
+
+export default stringConstants;
```
Comment (StanevichIvan) [inline src/constants/constants.ts]: let's avoin default exports

File: src/index.html [hunk 43]
```diff
@@ -6,8 +6,11 @@
   <base href="/">
   <meta name="viewport" content="width=device-width, initial-scale=1">
   <link rel="icon" type="image/x-icon" href="favicon.ico">
+  <link rel="preconnect" href="https://fonts.gstatic.com">
+  <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500&display=swap" rel="stylesheet">
```
Comment (StanevichIvan) [inline src/index.html]: Might we could place this links to the angular.json?

File: src/style/mixins.scss [hunk 44]
```diff
@@ -0,0 +1,32 @@
...
+  align-items: center;
+  justify-content: center;
+}
+
+@mixin icon-element {
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: center;
+}
```
Comment (StanevichIvan) [inline src/style/mixins.scss]: please add empty line

File: src/styles.scss [hunk 45]
```diff
@@ -1 +1,15 @@
-/* You can add global styles to this file, and also import other style files */
+@import "./style/variables.scss";
+@import "./style/mixins.scss";
+@import "./style/functions.scss";
```
Comment (StanevichIvan) [inline src/styles.scss:3]: Do we have reset scss?

**Review by StanevichIvan — DISMISSED** [review]

**Review by StanevichIvan — APPROVED** [review]

**Review by AnatoliRub — APPROVED** [review]

**Pull request merged**